keras_nlp/layers/transformer_encoder.py

```diff
@@ -0,0 +1,164 @@
+
+    Examples:
+
+    ```python
+    # Create a single transformer decoder layer.
+    encoder = keras_nlp.layer.TransformerEncoder(
+        feedforward_dim=64, num_heads=8)
+
+    # Create a simple model containing the decoder.
+    input = tf.keras.Input(shape=[4, 6])
+    output = encoder(input)
+    model = tf.keras.Model(inputs=input, outputs=output)
+
+    # Call decoder on the inputs.
+    input_data = tf.random.uniform(shape=[1, 10, 64])
+    output = model(input_data)
+
+    ```
+    """
+
+    def __init__(
+        self,
+        feedforward_dim,
+        num_heads,
+        dropout=0,
+        activation="relu",
+        layer_norm_epsilon=1e-05,
+        name="TransformerEncoder",
+        **kwargs
+    ):
+        super().__init__(name=name, **kwargs)
+        self._feedforward_dim = feedforward_dim
+        self._num_heads = num_heads
+        self._dropout = dropout
+        self._activation = activation
+        self._layer_norm_epsilon = layer_norm_epsilon
+        self._built = False
+
+    def _build(self, input_shape):
+        # Create layers based on input shape.
+        self._built = True
+        feature_size = input_shape[-1]
+        self._attention_head_size = int(feature_size // self._num_heads)
+        self._multi_head_attention_layer = keras.layers.MultiHeadAttention(
+            num_heads=self._num_heads,
+            key_dim=self._attention_head_size,
+            value_dim=self._attention_head_size,
+            dropout=self._dropout,
+        )
+
+        self._attention_layernorm = keras.layers.LayerNormalization()
+        self._feedforward_layernorm = keras.layers.LayerNormalization()
+
+        self._intermediate_dense = keras.layers.Dense(
+            self._feedforward_dim, activation=self._activation
+        )
+        self._intermediate_dropout = keras.layers.Dropout(rate=self._dropout)
+        self._output_dense = keras.layers.Dense(feature_size)
+        self._output_dropout = keras.layers.Dropout(rate=self._dropout)
+
+    def _add_and_norm(self, input1, input2, norm_layer):
+        return norm_layer(input1 + input2)
+
+    def _feed_forward(self, input):
+        x = self._intermediate_dense(input)
+        x = self._intermediate_dropout(x)
+        x = self._output_dense(x)
+        return self._output_dropout(x)
+
+    def call(self, inputs, padding_mask=None, additional_mask=None):
+        """Forward pass of the TransformerEncoder.
+
+        Args:
+            inputs: a Tensor. The input data to TransformerEncoder, should be
+                of shape [batch_size, sequence_length, feature_dim].
+            padding_mask: a boolean Tensor. It indicates if the token should be
+                masked because the token is introduced due to padding.
+                `padding_mask` should have shape [batch_size, sequence_length].
+                False means the certain certain is masked out.
+            additional_mask: a boolean Tensor. Customized mask used to mask out
```

Comment:
This should be `attention_mask`, since it masks attention locations. The new attention mask becomes the combination of this one and the padding mask.